Notify consensus of previously executed blocks in restart scenarios

georgepisaltu opened the pull request:
Fixes #3618

This PR adds a new announcement for finalized blocks enqueued for execution, but which contract runtime will not execute because their height is lower than the next expected executable height. In this scenario, contract runtime will raise this announcement which will be handled by the reactor. If there is a complete block in storage with that same height, it means the block can be safely considered executed and the reactor raises an event letting consensus know that the finalized block it had enqueued was handled, the same way it would happen if the block was executed. Consensus is oblivious to the fact that execution didn't take place.

We do this to ensure the smooth operation of the consensus validating process in the case of a restart in the middle of an era, where the restarted node couldn't make any consensus proposals until all the previous blocks in the era along with their vertices were "executed".

georgepisaltu committed:
Introduce `UnexecutedBlockAnnouncement`

This commit adds a new announcement for finalized blocks enqueued for
execution, but which contract runtime will not execute because their
height is lower than the next expected executable height. In this
scenario, contract runtime will raise this announcement which will be
handled by the reactor. If there is a complete block in storage with
that same height, it means the block can be safely considered executed
and the reactor raises an event letting consensus know that the
finalized block it had enqueued was handled, the same way it would
happen if the block was executed. Consensus is oblivious to the fact
that execution didn't take place.

We do this to ensure the smooth operation of the consensus validating
process in the case of a restart in the middle of an era, where the
restarted node couldn't make any consensus proposals until all the
previous blocks in the era along with their vertices were "executed".

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

rafal-ch commented on node/src/components/storage.rs:
```diff
@@ -1571,6 +1571,27 @@ impl Storage {
 
+    /// Retrieves a single complete block header by height by looking it up in
+    /// the index and returning it.
+    pub fn read_complete_block_header_by_height(
```
Would it be possible to use the already existing [get_block_header_by_height_restricted()](https://github.com/casper-network/casper-node/blob/a54d19ba1e4484d26961a1f48da3ba8bffa38322/node/src/components/storage.rs#L1948)?

georgepisaltu replied:
It's certainly possible, but I didn't want to do that mainly because `get_block_header_by_height_restricted` looks like a function which may change in the future (I'm talking about the restrictions) and I wanted something explicit here. I have no strong opinion on it though, we can safely use `get_block_header_by_height_restricted` if it's better.

georgepisaltu replied:
Forgot to mention in my previous comment, `get_block_header_by_height_restricted` also requires a transaction instance (which is probably why it's not a `pub fn`) and I don't want to create those outside of storage.

rafal-ch replied:
Yes, I agree there are differences. Despite not liking this code duplication much, I'm not strong about it. So let it be :)

Anyway, I also see that the only restriction we now have is `only_from_available_block_range`, which is set to `true` in all the flows. This is probably a remnant of the previous implementation of fast sync and we can as well get rid of it, hitting two birds with one stone. That'd be something to consider as a separate ticket (cc: @Fraser999)

Fraser999 replied:
Happy for that suggestion to get folded into this PR, or if not @georgepisaltu let me know and we can create a ticket request.

georgepisaltu replied:
We will need to create a separate ticket for this cleanup. I tried to do it but it's not as simple as it looks because "only from available block range" is not the same as "complete". There are multiple usages of these functions in storage and I believe they are due for a cleanup, but it would take more investment than it would make sense for the issue this PR solves.

rafal-ch commented on node/src/reactor/main_reactor.rs:
```diff
@@ -448,6 +448,42 @@ impl reactor::Reactor for MainReactor {
+                    // era by enqueuing all finalized blocks starting from the
+                    // first one in that era, blocks which should have already
+                    // been executed and marked complete in storage.
+                    warn!(
+                        "Finalized block with height {} enqueued for execution, \
+                        but a complete block header with the same height is \
+                        not present in storage.",
+                        block_height
+                    );
```
Nit:

Consider logging using this method:

```suggestion
                    warn!(
                        block_height,
                        "Finalized block enqueued for execution, but a complete \
                        block header with the same height is not present in storage."
                    );
```

georgepisaltu replied:
Will do.

Fraser999 replied:
I'd actually consider this an case worthy of an error level log.

georgepisaltu replied:
Done.

fizyk20 reviewed (approved):
Looks good :+1: (although I do like Rafał's suggestion about logging :wink:)

georgepisaltu committed:
Reformat log message

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Merge remote-tracking branch 'upstream/dev' into fix_itst14

georgepisaltu committed:
Fix get complete block header by height

This commit uses the `completed_blocks` index rather than the available
block range to fetch a complete block header by height in
`read_complete_block_header_by_height`.

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Fix lints and formatting

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Revert "Fix get complete block header by height"

This reverts commit cdf1da35bb7f411d0cc02b6db754f7666a9d225a.

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Update storage docs for reading complete headers

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu commented:
bors r+
